src/Framework/Models/EagerLoader.php

```diff
@@ -0,0 +1,127 @@
+<?php
+
+namespace Give\Framework\Models;
+
+use ReflectionClass;
+
+/**
+ * @unreleased
+ */
+class EagerLoader
+{
+    /**
+     * @unreleased
+     * @var string
+     */
+    protected $modelClass;
+
+    /**
+     * @unreleased
+     * @var string
+     */
+    protected $relationshipKey;
+
+    /**
+     * @unreleased
+     * @var ModelQueryBuilder
+     */
+    protected $modelQuery;
+
+    /**
+     * @unreleased
+     * @var ModelQueryBuilder
+     */
+    protected $eagerLoadedQuery;
+
+    /**
+     * @unreleased
+     * @var string
+     */
+    protected $foreignKey;
+
+    /**
+     * @var mixed
+     */
+    protected $foreignAttribute;
+
+    public function __construct($modelClass, $relationshipKey, $eagerLoadedModelClass, $foreignKey, $foreignAttribute = null)
+    {
+        $this->modelClass = $modelClass;
+        $this->modelQuery = $modelClass::query();
+        $this->relationshipKey = $relationshipKey;
+        $this->eagerLoadedQuery = $eagerLoadedModelClass::query();
+        $this->foreignKey = $foreignKey;
+        $this->foreignAttribute = $foreignAttribute ?? $foreignKey;
+    }
+
+    /**
+     * @unreleased
+     */
+    public function __call($name, $arguments)
+    {
+        $this->modelQuery->$name(...$arguments);
+        return $this;
+    }
+
+    /**
+     * @unreleased
+     *
+     * This method wraps the `get()` method of the underlying ModelQueryBuilder.
+     * It uses the results to query the related models and pre-set the cachedRelations property.
+     */
+    public function get()
```

Comment:
Can you make sure the IDE is aware of the return types on these? Jason did something fancy with generic in the model query builder that worked pretty well.  Basically each get and getAll can either return model instance(s) or null

Reply:
Yeah, I was able to replicate what Jason did.

Reply:
Resolved, see https://github.com/impress-org/givewp/pull/7227/commits/69707995eddb9f276cc69aa9c51272f72cc207e4.